sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+            try {
+                debugOut = new PrintWriter(new FileWriter("/home/florian/debug out"));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        debugOut.println(str);
+        debugOut.flush();
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadCreate extends LLVMBuiltin {
+        @Child
+        LLVMStoreNode store = null;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object thread, Object attr, Object startRoutine, Object arg) {
+            // create store node
+            if (store == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                store = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createStoreNode(LLVMInteropType.ValueKind.I64);
+            }
+
+            // create thread for execution of function
+            Thread t = LLVMLanguage.getLLVMContextReference().get().getEnv().createThread(() -> {
+                CompilerDirectives.transferToInterpreter();
+                // RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(getLLVMLanguage()));
+                RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(LLVMLanguage.getLanguage()));
+
+                // pthread_exit throws a control flow exception to stop the thread
+                try {
+                    callTarget.call(startRoutine, arg);
+                } catch (ControlFlowException e) {
+
+                }
+            });
+            // store cur id in thread var
+            store.executeWithTarget(thread, t.getId());
+
+            // store thread with thread id in context
+            LLVMLanguage.getLLVMContextReference().get().threadStorage.put(t.getId(), t);
+            // start thread
+            t.start();
+
+            return 0;
+        }
+    }
+
+    static final class MyArgNode extends LLVMExpressionNode {
+        private FrameSlot slot;
+
+        private MyArgNode(FrameSlot slot) {
+            this.slot = slot;
+        }
+
+        @Override
+        public Object executeGeneric(VirtualFrame frame) {
+            return frame.getValue(slot);
+        }
+    }
+
+    private static class RunNewThreadNode extends RootNode {
+        @Child
+        LLVMExpressionNode callNode = null;
+
+        @CompilerDirectives.CompilationFinal
+        FrameSlot functionSlot = null;
+
+        @CompilerDirectives.CompilationFinal
+        FrameSlot argSlot = null;
+
+        @CompilerDirectives.CompilationFinal
+        FrameSlot spSlot = null;
+
+        private LLVMLanguage language;
+
+        protected RunNewThreadNode(LLVMLanguage language) {
+            super(language);
```

Comment:
The `FrameDescriptor` can be a static field that is populated with the `FrameSlot`s you need in the static initializer. Then you can pass it to the `super` call as a second argument and initialize your slots and the `callNode` in this constructor.